src/Carousel/widget/Carousel.ts

```diff
@@ -0,0 +1,207 @@
+    }
+
+    public createProps(): ImageCarouselProps {
+        return {
+            callMicroflow: this.callMicroflow,
+            captionAttr: this.captionAttr,
+            contextId: this.contextObject ? this.contextObject.getGuid() : "",
+            data: this.data,
+            dataSourceMicroflow: this.dataSourceMicroflow,
+            descriptionAttr: this.descriptionAttr,
+            entityConstraint: this.entityConstraint,
+            height: this.height,
+            heightUnits: this.heightUnits,
+            imageEntity: this.imageEntity,
+            imageSource: this.imageSource,
+            interval: this.interval,
+            isLoading: this.isLoading,
+            onClickEvent: this.onClickEvent,
+            pageForm: this.pageForm,
+            pageLocation: this.pageLocation,
+            requiresContext: this.requiresContext,
+            staticImageCollection: this.staticImageCollection,
+            width: this.width,
+            widthUnits: this.widthUnits,
+        };
+    }
+    public postCreate() {
+        this.updateRendering();
+    }
+    /**
+     * called when context is changed or initialized. 
+     *
+     */
+    public update(object: mendix.lib.MxObject, callback: Function): void {
+        this.contextObject = object;
+        this.updateData(() => {
+            this.isLoading = false;
+            this.updateRendering(callback);
+        });
+        this.resetSubscriptions();
+    }
+    /**
+     * called when the widget is destroyed.
+     * will need to unmount react components
+     */
+    public uninitialize(): void {
+        unmountComponentAtNode(this.domNode);
+    }
+    /**
+     *  called to render the interface 
+     */
+    private updateRendering(callback?: Function) {
+        let props = this.createProps();
+        props.widgetId = this.id;
+        render(
+            createElement(ImageCarousel, props), this.domNode
+        );
+        if (callback) {
+            callback();
+        }
+    }
+    /**
+     * Determines which data source was specific and calls the respective method to get the Data 
+     */
+    private updateData(callback: Function): void {
+        if (this.imageSource === "xpath" && this.imageEntity) {
+            this.fetchDataFromXpath((objects) => {
+                this.setData(objects);
+                callback();
+            });
+        } else if (this.imageSource === "microflow" && this.dataSourceMicroflow) {
+            this.fetchDataFromMicroflow((objects) => {
+                this.setData(objects);
+                callback();
+            });
+        }
+        callback();
+    }
+    private fetchDataFromXpath(callback: (objects: mendix.lib.MxObject[]) => void): void {
+        const isMissingContext = (this.requiresContext && !this.contextObject) ||
+                                 (!this.requiresContext && this.entityConstraint.indexOf("[%CurrentObject%]") > -1 );
+        if (!isMissingContext) {
```

Comment:
This might be confusing, as `hasContext` is not equal to `!isMissingContext`  is it also requires that context is needed, else we don't care. Will need to split it `!needsContext || needConext && hasContext`